File: 二分查找/33.搜索旋转排序数组.py
# 二分法的精髓似乎就是：
# 1. **判断** 目标(要寻找的值) **在m的左边还是右边**
# 2. 有时(难题)需要借助其他值来判断 这个值也叫 pivot
# 3. 所以我们经常要用分类讨论
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1
        while l <= r:
            m = l + (r - l) // 2
            if target > nums[r]:
                if nums[m] > nums[r] and nums[m] < target:
                    l = m + 1
                elif nums[m] == target:
                    return m
                else:
                    r = m - 1
            else:
                if nums[m] > target and nums[m] < nums[r]:
                    r = m - 1
                elif nums[m] == target:
                    return m
                else:
                    l = m + 1
        return -1

File: 二分查找/33.搜索旋转排序数组.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var search = function(nums, target) {
    let l = 0, r = nums.length - 1, m;
    while (l <= r) {
        m = (l + r) >> 1;
        if (nums[m] === target) return m;
        if (nums[m] <= nums[r]) {
            if (target > nums[m] && target <= nums[r]) l = m + 1;
            else r = m - 1;
        } else {
            if (target < nums[m] && target >= nums[l]) r = m - 1;
            else l = m + 1;
        }
    }
    return -1;
};